pkg/dynamicaccess/controller.go

```diff
@@ -0,0 +1,283 @@
+	return &ControllerStruct{
+		accessLogic: accessLogic,
+	}
+}
+
+func (c *ControllerStruct) DownloadHandler(
+	ctx context.Context,
+	ls file.LoadSaver,
+	encryptedRef swarm.Address,
+	publisher *ecdsa.PublicKey,
+	historyRef swarm.Address,
+	timestamp int64,
+) (swarm.Address, error) {
+	_, act, err := c.getHistoryAndAct(ctx, ls, historyRef, publisher, timestamp)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+
+	return c.accessLogic.DecryptRef(ctx, act, encryptedRef, publisher)
+}
+
+func (c *ControllerStruct) UploadHandler(
+	ctx context.Context,
+	ls file.LoadSaver,
+	reference swarm.Address,
+	publisher *ecdsa.PublicKey,
+	historyRef swarm.Address,
+) (swarm.Address, swarm.Address, swarm.Address, error) {
+	history, act, err := c.getHistoryAndAct(ctx, ls, historyRef, publisher, time.Now().Unix())
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+	actRef := swarm.ZeroAddress
+	newHistoryRef := historyRef
+	if historyRef.IsZero() {
+		newHistoryRef, actRef, err = c.saveHistoryAndAct(ctx, history, nil, act)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+	}
+
+	encryptedRef, err := c.accessLogic.EncryptRef(ctx, act, publisher, reference)
+	return actRef, newHistoryRef, encryptedRef, err
+}
+
+// Limitation: If an upadate is called again within a second from the latest upload/update then mantaray save fails with ErrInvalidInput,
+// because the key (timestamp) is already present, hence a new fork is not created
+func (c *ControllerStruct) UpdateHandler(
+	ctx context.Context,
+	ls file.LoadSaver,
+	gls file.LoadSaver,
+	encryptedglRef swarm.Address,
+	historyRef swarm.Address,
+	publisher *ecdsa.PublicKey,
+	addList []*ecdsa.PublicKey,
+	removeList []*ecdsa.PublicKey,
+) (swarm.Address, swarm.Address, swarm.Address, swarm.Address, error) {
+	history, act, err := c.getHistoryAndAct(ctx, ls, historyRef, publisher, time.Now().Unix())
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+
+	gl, err := c.getGranteeList(ctx, gls, encryptedglRef, publisher)
+	if err != nil {
+		return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+	}
+	if len(addList) != 0 {
+		err = gl.Add(addList)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+	}
+	granteesToAdd := addList
+	if len(removeList) != 0 {
+		err = gl.Remove(removeList)
+		if err != nil {
+			return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+		}
+		// generate new access key and new act, only if history was not newly created
+		if !historyRef.IsZero() {
+			act, err = c.newActWithPublisher(ctx, ls, publisher)
+			if err != nil {
+				return swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, swarm.ZeroAddress, err
+			}
+		}
+		granteesToAdd = gl.Get()
+	}
+
+	for _, grantee := range granteesToAdd {
```

Comment:
Is there a way that maybe both `addList` and `removeList` might be `nil` or `empty`  and here we are trying to iterate over a `nil` or `empty` list ?

Reply:
and is that a problem?

Reply:
Empty and nil slice has 0 length, iterating over it does nothing. I don't think it is a problem.

Reply:
From the code perspective there should be no issues, just asking from the logic perspective.